data-showcase/grails-app/services/nl/thehyve/datashowcase/ProjectService.groovy

```diff
@@ -7,20 +7,95 @@
 package nl.thehyve.datashowcase
 
 import grails.gorm.transactions.Transactional
+import groovy.transform.CompileStatic
 import nl.thehyve.datashowcase.representation.ProjectRepresentation
+import nl.thehyve.datashowcase.representation.SearchQueryRepresentation
+import nl.thehyve.datashowcase.search.SearchCriteriaBuilder
+import org.grails.core.util.StopWatch
+import org.hibernate.Criteria
+import org.hibernate.HibernateException
+import org.hibernate.SessionFactory
+import org.hibernate.Transaction
+import org.hibernate.criterion.Criterion
+import org.hibernate.criterion.Projections
+import org.hibernate.criterion.Restrictions
 import org.modelmapper.ModelMapper
 import org.springframework.beans.factory.annotation.Autowired
 
 @Transactional(readOnly = true)
 class ProjectService {
 
     @Autowired
+    Environment dataShowcaseEnvironment
+
+    @Autowired
+    SearchCriteriaBuilder searchCriteriaBuilder
+
+    SessionFactory sessionFactory
+
+    @Autowired
     ModelMapper modelMapper
 
+    @CompileStatic
+    static ProjectRepresentation map(Map projectData) {
+        new ProjectRepresentation(
+                name: projectData.name as String,
+                lineOfResearch: projectData.lineOfResearch as String
+        )
+    }
+
     List<ProjectRepresentation> getProjects() {
         Project.findAll().collect({
             modelMapper.map(it, ProjectRepresentation)
         })
     }
 
+    List<ProjectRepresentation> getProjects(Set concepts, SearchQueryRepresentation searchQuery) {
+
+        Criterion searchQueryCriterion = searchQuery ? searchCriteriaBuilder.buildCriteria(searchQuery) : null
+        def stopWatch = new StopWatch('Fetch projects filtered items')
+        stopWatch.start('Retrieve from database')
+        // stateless session does not support collections of associated objects
+        // http://forum.spring.io/forum/spring-projects/batch/37785-collections-cannot-be-fetched-by-a-stateless-session
+        def session = sessionFactory.openSession()
+        Transaction tx = null
```

Comment:
But there is already `Transactional` on top of the class. Why is this needed?

Reply:
right, I changed a stateless session to the stateful one and I thought this is needed then. I'll remove it.